torch/_inductor/fx_passes/quantization.py

```diff
@@ -2395,6 +2395,56 @@ def _generate_qlinear_weight_prepack_patterns(
         )
 
 
+def _generate_linear_dynamic_fp16_pattern(
+    _dequant_weight_pattern,
+    input_dim_exceeds_two=False,
+):
+    dtype = torch.float32
+    t_pattern = _generate_linear_t_pattern(_dequant_weight_pattern, dtype)
+    x_q_dq_patten = CallFunction(
+        quantized_decomposed.dequantize_per_tensor.default,
+        CallFunction(
+            quantized_decomposed.quantize_per_tensor.default,
+            KeywordArg("x"),
+            KeywordArg("x_scale"),
+            KeywordArg("x_zp"),
+            KeywordArg("x_quant_min"),
+            KeywordArg("x_quant_max"),
+            KeywordArg("x_q_dtype"),
+        ),
+        Arg(),  # x_scale
+        Arg(),  # x_zp
+        Arg(),  # x_q_min
+        Arg(),  # x_q_max
+        Arg(),  # x_q_dtype
+    )
```

Comment:
Did we remove the pair of quant_fp16/dequant_fp16 at the activation?

Reply:
Yes. Computation is done in float32, so quant is not needed. We inserting quant/dequant here because technically we cannot insert for weight alone. Activation has an observer too.

Reply:
How was the fp16 linear computed in the legacy path? Did we ignore the fp32->fp16->fp32 too? The concern is that the data precision seems different between the FX graph presentation and the fused one.

Reply:
Yes. The dtype difference is real. In the graph after `convert_pt2e` before fusion, the pattern is `x -> to_fp16 -> to_fp32 -> linear`. But in the FX pass, the pattern is `x -> linear`. After fusion, the issue is gone.
The problem is that, with the PT2E pass, we cannot avoid inserting `d/dq` for activation. This op is more like weight-only quantization than dynamic quantization and the PT2E path does not work for that.
So, I guess maybe it's acceptable since users are supposed to do fusion before running.

Reply:
Just figured out how to remove `q/dq` for activation. Will modify this part. Thanks.